added support for two binds per direction

ohmailon opened the pull request:
In Fortnite you can bind a direction of movement to two different keys, so i added support for it in the wooting double movement utility too.
This is useful if you have bindings like me, where u can't press some key combinations while strafing / moving if u have only one key assigned.

![image](https://user-images.githubusercontent.com/57702380/116325882-819ddd00-a7c3-11eb-8406-2a87c88a89bd.png)

simon-wh commented on src/native/src/config.rs:
```diff
@@ -30,16 +34,24 @@ impl Default for KeyMapping {
             #[cfg(windows)]
             left_joystick: JoystickKeyMapping {
                 up: KeyId::to_u8(&KeyId::W).unwrap(),
+                up_two: KeyId::to_u8(&KeyId::W).unwrap(),
```
It kinda seems a little weird to have both the primary and secondary binds defined at the same time by default. Maybe we should change the bindings to be Option<u8>, then the secondary binds can be undefined by default

simon-wh commented on src/native/src/controller.rs:
```diff
@@ -8,17 +8,25 @@ use winapi::um::winuser::GetAsyncKeyState;
 #[derive(Debug)]
 pub struct JoystickState {
     up: bool,
+    up_two: bool,
```
The purpose of the JoystickState is to keep track of the state of each direction, so it shouldn't really have each direction duplicated. It would be better to update the directon_state based on something like ( bindOneState || bindTwoState )

simon-wh commented on src/native/src/controller.rs:
```diff
@@ -119,24 +135,28 @@ impl JoystickState {
     #[allow(dead_code)]
     pub fn update_key_states(&mut self, mappings: &JoystickKeyMapping) -> bool {
         self.update_key_state(JoystickDirection::Up, mappings.up)
+            | self.update_key_state(JoystickDirection::UpTwo, mappings.up_two)
```
Related to my other comment. It'd be better to allow `update_key_state` to accept multiple bindings for a particular direction and compute the result of those bindings, rather than have separated directions. This would also simplify the logic of the get_*_direction functions

simon-wh reviewed (changes requested):
I've just looked at the Rust side. It's pretty good, main things are that the secondary bindings shouldn't be defined by default & the secondary bindings should be abstracted out before it gets to JoystickState

simon-wh commented on src/native/src/service.rs:
```diff
@@ -160,18 +160,34 @@ impl Service {
                         .controller_state
                         .left_joystick
                         .set_direction_state(JoystickDirection::Up, state == State::Pressed),
+                    x if x == self.config.key_mapping.left_joystick.up_two => self
```
The suggestion to abstract the double bindings out of the JoystickState makes the rawinput method a little tricky, as it works based on Key Pressed / Released events. Although, we aren't actively using the rawinput implementation anymore, but it would be nice to keep it fully working

ohmailon replied:
done with all requested changes :)

ohmailon commented:
Hi, i implemented your requested changes. I also managed to get the raw input to work by adding a custom keybind state if raw input is enabled. 

I'll now merge those change into the https://github.com/WootingKb/wooting-double-movement/pull/18 branch

ohmailon commented:
Hi, i implemented your requested changes. I also managed to get the raw input to work by adding a custom keybind state if raw input is enabled.

I'll now merge those change into the #18 branch

ohmailon commented:
Hi, i implemented your requested changes. I also managed to get the raw input to work by adding a custom keybind state if raw input is enabled. 

I also merged these changes into the #18 branch

simon-wh commented:
Great! I hope you're enjoying your first adventure into this stack of technologies, you appear to be taking it in stride, good work!

simon-wh commented on src/native/src/config.rs:
```diff
@@ -30,16 +34,24 @@ impl Default for KeyMapping {
             #[cfg(not(windows))]
             left_joystick: JoystickKeyMapping {
                 up: 0,
+                up_two: 0,
```
These need updated to reflect the updated types of these properties `None`

simon-wh commented on src/native/src/controller.rs:
```diff
@@ -96,32 +97,33 @@ impl JoystickState {
     }
 
-    pub fn set_direction_state(&mut self, direction: JoystickDirection, state: bool) -> bool {
+    pub fn set_direction_state(&mut self, direction: JoystickDirection, state: bool, state_two: bool) -> bool {
```
I would maybe move the `state_two` out of here and just pass the final result in the `update_key_state` function

simon-wh commented on src/native/src/controller.rs:
```diff
@@ -96,32 +97,33 @@ impl JoystickState {
 
     #[allow(dead_code)]
-    pub fn update_key_state(&mut self, direction: JoystickDirection, binding: u8) -> bool {
+    pub fn update_key_state(&mut self, direction: JoystickDirection, bind_one: u8, bind_two: u8) -> bool {
```
For the `bind_two` you could probably keep passing it as `Option<u8>` and not unwrap it in the call to this function.

Something like this would be nice:
```
update_key_state(..., bind_two: Option<&u8>) (for this you'll need to pass the bind using .as_ref())

let mut key_state = get_key_state(bind_one);
if let Some(bind_two) = bind_two {
   key_state |= get_key_state(bind_two);
}
self.set_direction_state(direction, key_state);
```
The calls to GetAsyncKeyState and processing that result into a nice bool could probably be moved to its own helper function to save repeating the logic for it. (Just remember to slap a wee `#[cfg(windows)]` on that function so that the project can be compiled on Mac/Linux for UX development)

simon-wh commented:
I like the choice of using `Esc` to unbind a key, I think it'd also be good to have the option to right-click the box as well to remove the binding.

PastaJ36 commented on src/native/types.ts:
```diff
@@ -24,11 +28,16 @@ export interface KeyMapping {
 export const defaultKeyMapping: KeyMapping = {
   leftJoystick: {
     up: Key.W,
+    up_two: undefined,
     down: Key.S,
+    down_two: undefined,
     left: Key.A,
+    left_two: undefined,
     right: Key.D,
+    right_two: undefined,
```
These undefined defaults can just be left out

PastaJ36 commented on src/view/AdvancedSettings.tsx:
```diff
@@ -120,21 +108,30 @@ function AngleControl() {
       (event) => {
-        console.log(event);
-        props.valueChanged(event.keyCode);
+        props.valueChanged(event.keyCode === Key.Escape && optional ? undefined : event.keyCode);
```
I think it's better to use right-click to unbind then this escape exception. Now it's also not possible to bind Escape to the second binds. We should also apply the unbinding behaviour to both keysbinds, make them the same.

ohmailon commented:
Requested changes are implemented + you can now unbind a key by `right clicking` the assigned field OR by clicking the input field and pressing `Esc`

> Great! I hope you're enjoying your first adventure into this stack of technologies, you appear to be taking it in stride, good work!

Yes definitely 😃
Rust may take some time getting used to but its it but it's an interesting challenge.
React on the other hand has a lot in common with Angular in the way it works (like the component system & value stores) , its just implemented in a different way.

PastaJ36 commented on src/view/AdvancedSettings.tsx:
```diff
@@ -194,30 +171,86 @@ export function KeyBinding() {
-          value={keyMapping.leftJoystick.right}
-          valueChanged={(value) => assignNewJoystickBind("right", value)}
-        />
+        <Flex>
```
This part is getting quite long now. We can split up all the components in different files.

PastaJ36 reviewed (commented):
Looks pretty good, just some small things. Last thing is that we now use prettier as a default formatter. We haven't set it up with a nice CI yet, but can you format your commit with prettier as well?

ohmailon commented:
All keybinds are now optional + `Esc` can be binded as any other key now & i let prettier do some cleanups  :)

simon-wh commented:
Looking really good, great work! I'll just play around with the app for a bit to make sure I haven't missed anything. 

Before merging I think we'll need to merge the `fix/ds4` branch so that this can be merged into main

ohmailon commented:
I've used it in Fortnite now for about an hour and it worked everything like it should :)

ohmailon commented:
Changed the following issues
- fixed bug wich allowed to assign multiple keys at once 
- added remove icon to delete the binds for an entire row
- added the same key assign logic like in Fortnite since we are also adjusting the way that unbinding works (every bind only once per column)

simon-wh commented:
Great! I've merged the `fix/ds4` branch (even though there are a couple of bugs that could be fixed) so that this can be merged into main. Rebasing this branch on main would be pretty nice for cleaning up the git history but should be aight either way

simon-wh commented on src/view/AdvancedSettings.tsx:
```diff
@@ -120,104 +118,217 @@ function AngleControl() {
+  const editingStateState = useState<keyof JoystickKeyMapping>();
+  const [editingState, setEditingState] = editingStateState;
+
+  const listener = (event: any) => {
```
Instead of having this listener always running, you could use the [useEffect hook](https://reactjs.org/docs/hooks-effect.html#tip-optimizing-performance-by-skipping-effects) 

This means you can have a function that runs when the `editingState` value changes, i.e.:
```
useEffect(() => {
    const subscription = window.addEventListener(...)
    return () => {
       // do unsubscribe here as this will be called when the editingState changes so you can do cleanup based on the previous value of editingState
    }
}, [editingState])
```

ohmailon replied:
i didn't implemented it exactly as in your example because addEventListener doesn't return an observable but it should work in the same way :)

simon-wh commented on src/view/AdvancedSettings.tsx:
```diff
@@ -120,104 +118,217 @@ function AngleControl() {
+  editingState: [
+    keyof JoystickKeyMapping | undefined,
+    Dispatch<SetStateAction<keyof JoystickKeyMapping | undefined>>
+  ];
```
I feel like instead of passing down this state variable that instead you could do:

```
requestEdit: (keybind: keyof JoystickKeyMapping) => void
isEditing: boolean
```

then let the parent of this determine how edits should be managed

simon-wh commented on src/view/AdvancedSettings.tsx:
```diff
@@ -120,104 +118,217 @@ function AngleControl() {
+    Object.keys(keyMapping.leftJoystick).forEach((existingKey) => {
+      const isMainBind = !existingKey.endsWith("two");
+      if (
+        isMainBind === isNewMainBind &&
```
Am I right in taking this code as making sure the same bind can't appear twice in the same column? e.g. W key can only appear once in the primary bindings

Would be good to add some comments to detail what's going on here as it took me a bit to understand what the purpose of this was

ohmailon replied:
added some comments

simon-wh reviewed (commented):
I've added a few comments, mainly around the code for preventing multiple keybinds being edited at once

ohmailon commented:
Made the requested changes and added some comments

ohmailon commented:
I updated my fork, merged the main branch into my PR branch. Should now match your main branch.

simon-wh commented on src/view/AdvancedSettings.tsx:
```diff
@@ -120,71 +118,120 @@ function AngleControl() {
+        }
+      );
+    }
+  }, [editingState]);
```
Only thing I have is here that if you start a keybind edit and then another one you'll end up with two event listeners active listening to "keydown". So the function inside useEffect should also have 
```
return () => { 
    window.removeEventListener("keydown", listener);
}`
``

to cleanup any dangling listeners before adding a new one

simon-wh commented:
Awesome, only have one more thing on my side then I think it should be good to go. Not sure if @PastaJ36 wants to give it another go over before merging

ohmailon commented:
fixed it

simon-wh commented:
I was just doing some final sanity checks to make sure everything is all good to merge. I noticed that right-clicking the keybind box no longer does anything, wasn't that interaction intended to remove the keybind that was clicked on? Starting an edit and clicking Esc cancels the rebind, which is what is intended?

simon-wh commented:
Ah, okay, I reckon we could prob merge this then, finally :sweat_smile:

Great work again btw
